core/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -605,40 +630,165 @@ private void maybeFSyncTranslogs() {
+                        } catch (EngineClosedException | AlreadyClosedException ex) {
+                            // fine - continue;
+                        }
+                        continue;
+                    default:
+                        throw new IllegalStateException("unknown state: " + shard.state());
+                }
+            }
+        }
+    }
 
-    /**
-     * FSyncs the translog for all shards of this index in a defined interval.
-     */
-    final static class AsyncTranslogFSync implements Runnable {
-        private final IndexService indexService;
-        private final ThreadPool threadPool;
+    static abstract class BaseAsyncTask implements Runnable, Closeable {
+        protected final IndexService indexService;
+        protected final ThreadPool threadPool;
+        private final TimeValue interval;
+        private ScheduledFuture<?> scheduledFuture;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+        private volatile Exception lastThrownException;
 
-        AsyncTranslogFSync(IndexService indexService, ThreadPool threadPool) {
+        BaseAsyncTask(IndexService indexService, ThreadPool threadPool, TimeValue interval) {
             this.indexService = indexService;
             this.threadPool = threadPool;
+            this.interval = interval;
+            onTaskCompletion();
         }
 
-        boolean mustRun() {
+        boolean mustReschedule() {
             // don't re-schedule if its closed or if we dont' have a single shard here..., we are done
-            return (indexService.closed.get() || indexService.shards.isEmpty()) == false;
+            return (indexService.closed.get() || indexService.shards.isEmpty()) == false
+                && closed.get() == false && interval.millis() > 0;
         }
 
-        void schedule() {
-            threadPool.schedule(indexService.getIndexSettings().getTranslogSyncInterval(), ThreadPool.Names.SAME, AsyncTranslogFSync.this);
+        private synchronized void onTaskCompletion() {
+            if (mustReschedule()) {
+                indexService.logger.debug("scheduling {} every {}", toString(), interval);
+                this.scheduledFuture = threadPool.schedule(interval, getThreadPool(), BaseAsyncTask.this);
+            } else {
+                indexService.logger.debug("scheduled {} disabled", toString());
+                this.scheduledFuture = null;
+            }
         }
 
-        @Override
-        public void run() {
-            if (mustRun()) {
-                threadPool.executor(ThreadPool.Names.FLUSH).execute(() -> {
-                    indexService.maybeFSyncTranslogs();
-                    if (mustRun()) {
-                        schedule();
+        public final void run() {
+            try {
+                runInternal();
+            } catch (Exception ex) {
+                if (lastThrownException == null || sameException(lastThrownException, ex) == false) {
```

Comment:
This is neat ... I hope (later!) we can somehow make this generic, fold it into logger layer or something.